Phase 1 :- Data set up and environment

In [ ]:
import pandas as pd#read csv files in data frames
from sqlalchemy import create_engine, text#make bridge between database (postgresql)and pyhton

# ── Connection ─────────────────────────────────────────────────
engine = create_engine('postgresql://postgres:1234@localhost:5432/olist_db')
#engine as a phoneline between python and postgresql
#postgres is default postgresql username 
DATA_PATH = "J:/python/DA_project/Olist_Ecommerce_intelligence/data/raw/"

files = {
    'customers':                    'olist_customers_dataset.csv',
    'sellers':                      'olist_sellers_dataset.csv',
    'product_category_translation': 'product_category_name_translation.csv',
    'products':                     'olist_products_dataset.csv',
    'geolocation':                  'olist_geolocation_dataset.csv',
    'orders':                       'olist_orders_dataset.csv',
    'order_items':                  'olist_order_items_dataset.csv',
    'order_payments':               'olist_order_payments_dataset.csv',
    'order_reviews':                'olist_order_reviews_dataset.csv',
}

# ── Step 1: Drop all tables  in reverse order (children first) ──
# This avoids the foreign key conflict
drop_order = [
    'order_reviews',
    'order_payments',
    'order_items',
    'orders',
    'geolocation',
    'products',
    'product_category_translation',
    'sellers',
    'customers',
]

print("Dropping existing tables...")
with engine.connect() as conn:
    for table in drop_order:
        conn.execute(text(f'DROP TABLE IF EXISTS "{table}" CASCADE'))
        print(f"  Dropped: {table}")
    conn.commit()
print("All old tables dropped.\n")

# ── Step 2: Load CSVs in correct order (parents first) ──────────
load_order = [
    'customers',
    'sellers',
    'product_category_translation',
    'products',
    'geolocation',
    'orders',
    'order_items',
    'order_payments',
    'order_reviews',
]

for table in load_order:
    filename = files[table]
    print(f"Loading {filename} -> table: {table} ...", end=" ")
    df = pd.read_csv(DATA_PATH + filename)
    df.to_sql(table, engine, if_exists='replace', index=False)
    print(f"Done. ({len(df):,} rows)")

print("\nAll tables loaded successfully!")

Dropping existing tables...
  Dropped: order_reviews
  Dropped: order_payments
  Dropped: order_items
  Dropped: orders
  Dropped: geolocation
  Dropped: products
  Dropped: product_category_translation
  Dropped: sellers
  Dropped: customers
All old tables dropped.

Loading olist_customers_dataset.csv -> table: customers ... Done. (99,441 rows)
Loading olist_sellers_dataset.csv -> table: sellers ... Done. (3,095 rows)
Loading product_category_name_translation.csv -> table: product_category_translation ... Done. (71 rows)
Loading olist_products_dataset.csv -> table: products ... Done. (32,951 rows)
Loading olist_geolocation_dataset.csv -> table: geolocation ... Done. (1,000,163 rows)
Loading olist_orders_dataset.csv -> table: orders ... Done. (99,441 rows)
Loading olist_order_items_dataset.csv -> table: order_items ... Done. (112,650 rows)
Loading olist_order_payments_dataset.csv -> table: order_payments ... Done. (103,886 rows)
Loading olist_order_reviews_dataset.csv -> table: order_r

Phase 2 — Data Cleaning & Feature Engineering

We'll work through this in 4 sections inside our notebooks/01_data_cleaning.ipynb.

Section 1 — Load & explore the raw data:- Always explore before we clean — never clean blindly.

In [ ]:
#load all data into dataframes
import pandas as pd
from sqlalchemy import create_engine, text

# ── Connection ─────────────────────────────────────────────────
engine = create_engine('postgresql://postgres:1234@localhost:5432/olist_db')
customers=pd.read_sql('select * from customers',engine)
sellers=pd.read_sql('select * from sellers',engine)
products=pd.read_sql('select * from products',engine)
orders=pd.read_sql('select * from orders',engine)
order_items=pd.read_sql('select * from order_items',engine)
order_payments=pd.read_sql('select * from order_payments',engine)
order_reviews=pd.read_sql('select * from order_reviews',engine)
geoloacation=pd.read_sql('select * from geolocation',engine)
product_category_translation=pd.read_sql('select * from product_category_translation',engine)
print('All Tables Loaded Successfully')

All Tables Loaded Successfully


In [3]:
#First look at the orders dataframe
print("Shape:",orders.shape)
print('\nColumn Names:\n',orders.columns.to_list())
print("\nData Types:\n",orders.dtypes)
print("\nMissing Values:\n",orders.isnull().sum())
print("\nSample rows:")
orders.head()

Shape: (99441, 8)

Column Names:
 ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

Data Types:
 order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

Missing Values:
 order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Sample rows:


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


As we can see all valuable information from orders dataframe :

1.it has 99441 rows with 8 columns.

2.all the column names/fields of orders dataframe are well written in lower case with no space.

3.All data types of all attributes are object where as we need to change the date datatype to datetime further.

4.we have three columns where misiing values are non zero (order_approved_at,order_delivered_carrier_date,

order_delivered_customer_date)

5.orders.head():-shows our data with top 5 rows entry.

Below we shall do for all remaining dataframes

In [4]:
#First look at the order_items dataframe
print("Shape:",order_items.shape)
print('\nColumn Names:\n',order_items.columns.to_list())
print("\nData Types:\n",order_items.dtypes)
print("\nMissing Values:\n",order_items.isnull().sum())
print("\nSample rows:")
order_items.head()

Shape: (112650, 7)

Column Names:
 ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

Data Types:
 order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

Missing Values:
 order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Sample rows:


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


From order_items dataframe

1.shipping_limit_date is boject data type we need to change this data type to datetime

2.there is no missing values ,so we do not need to handle missing values

In [5]:
#First look at the customers dataframe
print("Shape:",customers.shape)
print('\nColumn Names:\n',customers.columns.to_list())
print("\nData Types:\n",customers.dtypes)
print("\nMissing Values:\n",customers.isnull().sum())
print("\nSample rows:")
customers.head()

Shape: (99441, 5)

Column Names:
 ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

Data Types:
 customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

Missing Values:
 customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Sample rows:


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


No Need to do anything in customers dataframe :)

In [6]:
#First look at the sellers dataframe
print("Shape:",sellers.shape)
print('\nColumn Names:\n',sellers.columns.to_list())
print("\nData Types:\n",sellers.dtypes)
print("\nMissing Values:\n",sellers.isnull().sum())
print("\nSample rows:")
sellers.head()

Shape: (3095, 4)

Column Names:
 ['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']

Data Types:
 seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

Missing Values:
 seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

Sample rows:


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


Everything is fine in Sellers dataframe as well

In [7]:
#First look at the products dataframe
print("Shape:",products.shape)
print('\nColumn Names:\n',products.columns.to_list())
print("\nData Types:\n",products.dtypes)
print("\nMissing Values:\n",products.isnull().sum())
print("\nSample rows:")
products.head()

Shape: (32951, 9)

Column Names:
 ['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

Data Types:
 product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

Missing Values:
 product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Sample rows:


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In Products dataframe as we can see mainly product_category_name has 610 missing values so we need to handle these as well

In [8]:
#First look at the order_payments dataframe
print("Shape:",order_payments.shape)
print('\nColumn Names:\n',order_payments.columns.to_list())
print("\nData Types:\n",order_payments.dtypes)
print("\nMissing Values:\n",order_payments.isnull().sum())
print("\nSample rows:")
order_payments.head()

Shape: (103886, 5)

Column Names:
 ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

Data Types:
 order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

Missing Values:
 order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Sample rows:


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


order_payments dataframe has all attributes in well written,no missing values.No need to change.

In [9]:
#First look at the order_reviews dataframe
print("Shape:",order_reviews.shape)
print('\nColumn Names:\n',order_reviews.columns.to_list())
print("\nData Types:\n",order_reviews.dtypes)
print("\nMissing Values:\n",order_reviews.isnull().sum())
print("\nSample rows:")
order_reviews.head()

Shape: (99224, 7)

Column Names:
 ['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

Data Types:
 review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object



Missing Values:
 review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Sample rows:


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,None,None,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,None,None,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,None,None,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,None,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,None,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In order_reviews dataframe there are lots of missing values in review_comment_title,review_comment_message.But there are no missing values in review_score.it is good otherwise we need to handle these missing values mainly because review_score of any product given by customer is mendatory.But for review_comment_title,review_comment_message we can not handle missing values because these are conversation between customer and the seller.In order_reviews dataframe two columns(review_creation_date,review_answer_timestamp) has object datatype we need to convert these into datetime[ns]

In [10]:
#name of geolocation in dataframe has changed  by mistake we need to change from geoloacation to geolocation
geolocation=geoloacation.copy()

In [11]:
#First look at the geolocation dataframe
print("Shape:",geolocation.shape)
print('\nColumn Names:\n',geolocation.columns.to_list())
print("\nData Types:\n",geolocation.dtypes)
print("\nMissing Values:\n",geolocation.isnull().sum())
print("\nSample rows:")
geolocation.head()

Shape: (1000163, 5)

Column Names:
 ['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']

Data Types:
 geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

Missing Values:
 geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

Sample rows:


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


No need to change in geolocation dataframe.all things are good

In [12]:
#First look at the product_category_translation dataframe
print("Shape:",product_category_translation.shape)
print('\nColumn Names:\n',product_category_translation.columns.to_list())
print("\nData Types:\n",product_category_translation.dtypes)
print("\nMissing Values:\n",product_category_translation.isnull().sum())
print("\nSample rows:")
product_category_translation.head()

Shape: (71, 2)

Column Names:
 ['product_category_name', 'product_category_name_english']

Data Types:
 product_category_name            object
product_category_name_english    object
dtype: object

Missing Values:
 product_category_name            0
product_category_name_english    0
dtype: int64

Sample rows:


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


No need to change anything in product_category_translation.

Section 2 :-Fix Data Types(Dates)

In [13]:
#for orders data frame
date_cols=[
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]
for cols in date_cols:
    orders[cols]=pd.to_datetime(orders[cols])
print(orders[date_cols].dtypes)

order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object


In [14]:
#for order_items data types(date)
order_items['shipping_limit_date']=pd.to_datetime(order_items['shipping_limit_date'])
#for order_reviews
date_cols=[
    'review_creation_date',
    'review_answer_timestamp'
]
for cols in date_cols:
    order_reviews[cols]=pd.to_datetime(order_reviews[cols])
print(order_reviews[date_cols].dtypes)
print(order_items['shipping_limit_date'].dtypes)

review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object
datetime64[ns]


Section3:-Check missing values for all Tables

In [15]:
for name,df in[('orders',orders),('products',products),('order_reviews',order_reviews),('customers',customers)]:
    missing=df.isnull().sum()
    missing=missing[missing>0]
    if len(missing)>0:
        print(f"\n{name}:")
        print(missing)
    else:
        print(f'\n{name}:NO Missing Values')



orders:
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64

products:
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

order_reviews:
review_comment_title      87656
review_comment_message    58247
dtype: int64

customers:NO Missing Values


Handling missing Values:


In [16]:
#1.orders some deliveries are missing (cancelled orders)
#We do not drop these rows.We keep them and handle in feature engineering.
print('order_status breakdown:')
print(orders['order_status'].value_counts())

#2.In products table some items missing category name
#Fill wiyh unknown so join's does not break
products['product_category_name']=products['product_category_name'].fillna('unknown')

#3.Order_reviews table missing data in review_comment_title and review_comment_message
#Fill with empty string so text operation's don't error
order_reviews['review_comment_title']=order_reviews['review_comment_title'].fillna('')
order_reviews['review_comment_message']=order_reviews['review_comment_message'].fillna('')
print('Missing values handled')

order_status breakdown:
order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64
Missing values handled


Section 4:-Feature Engineering:
    We create new columns from existing data that tell a richer buisness story.
    We'll engineer 8 new features. Each one directly feeds a KPI or analysis in later phases.

In [17]:
#Feature 1:-delivery_delay_days
#How many days late(or early) was the delivery?
#Positive=Late,Negative=Early,0=on time
#Buisness use:-Identify problem sellers by regions
orders['delivery_delay_days']=(
    orders['order_delivered_customer_date']-orders['order_estimated_delivery_date']
).dt.days
print("Delivery delay Stats:")
print(orders['delivery_delay_days'].describe())

#==================================================
#Feature 2:-is_late
#Boolean Flag-Was this order delivered late?
#Buisness Use:-Late delivery rate KPI
orders['is_late']=orders['delivery_delay_days']>0
print(f"\nLate Deliveries:{orders['is_late'].sum():,}")
print(f"Late Delivery Rate:{orders['is_late'].mean()*100:.1f}%")

#===================================================
#Feature 3:shipping_processing_days
#Days from purchase -> carrier pickup
#Buisness Use:-Seller fulfillment speed
orders['shipping_processing_days']=(
    orders['order_delivered_carrier_date']-orders['order_purchase_timestamp']
).dt.days
print("Shipping Stats:")
print(orders['shipping_processing_days'].describe())

#====================================================
#Feature 4 :- order_year,order_month,order_quarter
#Time-Based grouping for trend analysis
#Buisness Use:-MOM,YOY growth ,Seasonality
orders['order_year']=orders['order_purchase_timestamp'].dt.year
orders['order_month']=orders['order_purchase_timestamp'].dt.month
orders['order_quarter']=orders['order_purchase_timestamp'].dt.quarter
orders['order_yearmonth']=orders['order_purchase_timestamp'].dt.to_period('M')
print("\nOrder Date Range:")
print(orders['order_purchase_timestamp'].min(), "to" ,orders['order_purchase_timestamp'].max())


Delivery delay Stats:
count    96476.000000
mean       -11.876881
std         10.183854
min       -147.000000
25%        -17.000000
50%        -12.000000
75%         -7.000000
max        188.000000
Name: delivery_delay_days, dtype: float64

Late Deliveries:6,535
Late Delivery Rate:6.6%
Shipping Stats:
count    97658.000000
mean         2.743114
std          3.625481
min       -172.000000
25%          1.000000
50%          2.000000
75%          4.000000
max        125.000000
Name: shipping_processing_days, dtype: float64

Order Date Range:
2016-09-04 21:15:19 to 2018-10-17 17:30:18


Now  we Build a masters table/fact_table by joining everything together:

In [18]:
#=====================================================
#Build The Master table
#Join orders + order_items + products + product_category_translation +customers
#This is our main analytical dataset for all future work
#======================================================
#Step 1 Add category english names to products
products_full=products.merge(product_category_translation,on='product_category_name',how='left')

#Step 2:-Join order_items with products
items_full=order_items.merge(
    products_full[['product_id','product_category_name','product_category_name_english']],
    on='product_id',
    how='left'
    )
#Step3 Add payment info(sum per order - one order can have multiple payments)
payments_agg=order_payments.groupby('order_id').agg(
    total_payment=('payment_value','sum'),
    payment_type=('payment_type','first'),
    num_installments=('payment_installments','max')
).reset_index()
#Step 4 Add review Scores(one review per order)
reviews_clean=order_reviews[['order_id','review_score']].drop_duplicates('order_id')

#Step 5 :-Build master table
master=(
    orders
    .merge(customers[['customer_id','customer_unique_id','customer_city','customer_state']],
        on='customer_id',
        how='left')
    .merge(items_full,on='order_id',how='left')
    .merge(payments_agg,on='order_id',how='left')
    .merge(reviews_clean,on='order_id',how='left')
    .merge(sellers[['seller_id','seller_state']],on='seller_id',how='left')

)
print("MASTER Table Shape:",master.shape)
print("Columns:",master.columns)



MASTER Table Shape: (113425, 31)
Columns: Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'delivery_delay_days', 'is_late', 'shipping_processing_days',
       'order_year', 'order_month', 'order_quarter', 'order_yearmonth',
       'customer_unique_id', 'customer_city', 'customer_state',
       'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date',
       'price', 'freight_value', 'product_category_name',
       'product_category_name_english', 'total_payment', 'payment_type',
       'num_installments', 'review_score', 'seller_state'],
      dtype='object')


Now engineer 4 more new features to the masters table:

In [29]:
#Feature 5: revenue_per_item
#Actual revenue=price + freight
master['revenue_per_item']=master['price']+master['freight_value']

#Feature 6:frieght_ratio
#What % of item price is the shipping cost?
#High ratio=expensive to ship ->impact profitability
master['freight_ratio']=master['freight_value']/master['price'].replace(0,np.nan)  #to skip when price=0 then it replace to nan

#Feature 7: review_score_bucket
#Group 1-5 starts into readable buckets for dashboards
def score_bucket(score):
    if pd.isna(score): return 'No Review'
    elif score>=4:     return 'Positive (4-5)'
    elif score==3:      return 'Neutral (3)'
    else:              return 'Negative (1-2)'
master['review_bucket']=master['review_score'].apply(score_bucket)


In [45]:
#Feature 8:customer_order_number
#Which order number is this for that customer?
#1=First order 2=Repeat,etc.
#Buisness Use:-Repeat purchase rate,loyality analysis,customer retention,customer churn
master=master.sort_values('order_purchase_timestamp')
master['customer_order_number']=(
    master.groupby('customer_unique_id').cumcount()+1   #cumcount() start count with 0 so we do +1
)
master['is_repeat_customer']=master['customer_order_number']>1
repeat_rate=master.drop_duplicates('customer_unique_id')['is_repeat_customer'].mean()
master.shape
print(f"\nRepeat Customer Rate:{repeat_rate*100:.1f}%")


Repeat Customer Rate:0.0%


Our master table is at item level,not order level.One order with 3 products=3 rows,all with same customer_unique_id and order_id
customer_unique_id    order_id    product
abc123                order_001   shoes       ← row 1
abc123                order_001   shirt       ← row 2  (same order!)
abc123                order_001   belt        ← row 3  (same order!)
abc123                order_002   watch       ← row 4  (second order)

Why groupby('customer_unique_id').cumcount()+1   ->cumcount() count every row ,not every order
gives same order 3items=3 rows
drop_duplicates('customer_unique_id') ->keeps only first row:-it sees per customer
abc123                order_001   shoes       ← row 1
customer_order_number=1
is_repeat=Flase always for every customer

So the correct way is first we remove duplicates  drop_duplicates('order_id') first collapse the table to one row per order(removes the item level duplicates)
abc123                order_001   
abc123                order_002
then groupby -> max()  finds the highest order of each customer ever reached.
abc123                order_002
max order_number=2
is_repeat=True

In [62]:
# ── Correct way to calculate repeat rate ─────────────────────
# Get the highest order number each customer ever reached
customer_summary = (
    master
    .drop_duplicates('order_id')[['customer_unique_id', 'customer_order_number']]
    .groupby('customer_unique_id')['customer_order_number']
    .max()
    .reset_index()
)

customer_summary['is_repeat_customer'] = customer_summary['customer_order_number'] > 1

repeat_rate = customer_summary['is_repeat_customer'].mean()
print(f"Repeat Customer Rate: {repeat_rate*100:.1f}%")
print(f"Total unique customers: {len(customer_summary):,}")
print(f"Repeat customers: {customer_summary['is_repeat_customer'].sum():,}")
print(f"One-time customers: {(~customer_summary['is_repeat_customer']).sum():,}")

Repeat Customer Rate: 3.1%
Total unique customers: 96,096
Repeat customers: 2,997
One-time customers: 93,099


Section 5:-Save the cleaned data

PostgreSQL has no Period type — it is a pandas-only object. We must convert it to string before saving

In [69]:
master['order_yearmonth']=master['order_yearmonth'].astype(str)
print(master['order_yearmonth'].dtype)

object


In [ ]:
#To PostgreSQL - Phase 3 SQl analysis will query this
master.to_sql('master_orders',engine,if_exists='replace',index=False)
print('masters_orders saved to PostgreSQl')

#To CSV - backup + Tableau will use it directly
master.to_csv(r'J:\python\DA_project\Olist_Ecommerce_intelligence\data\processed\master_orders.csv',index=False)
print(r'master_orders.csv saved to Olist_Ecommerce_intelligence\data\processed')

masters_orders saved to PostgreSQl
master_orders.csv saved to Olist_Ecommerce_intelligence\data\processed


We also save customer_summary table seprately -It will be directly use in phase 4,5:

In [72]:
#save customer_summary for later phases
customer_summary.to_csv(r'J:\python\DA_project\Olist_Ecommerce_intelligence\data\processed\customer_summary.csv',index=False)
print("customer_summary saved:", customer_summary.shape)

customer_summary saved: (96096, 3)


In [74]:
print(f"""
══════════════════════════════════
  Phase 2 Complete — Summary
══════════════════════════════════
  Master table rows  : {len(master):,}
  Master table cols  : {master.shape[1]}

  New features created:
  - delivery_delay_days
  - is_late
  - shipping_processing_days
  - order_year / month / quarter
  - revenue_per_item
  - freight_ratio
  - review_bucket
  - customer_order_number
  - is_repeat_customer
══════════════════════════════════
""")


══════════════════════════════════
  Phase 2 Complete — Summary
══════════════════════════════════
  Master table rows  : 113,425
  Master table cols  : 36

  New features created:
  - delivery_delay_days
  - is_late
  - shipping_processing_days
  - order_year / month / quarter
  - revenue_per_item
  - freight_ratio
  - review_bucket
  - customer_order_number
  - is_repeat_customer
══════════════════════════════════



Before committing ,always validate.Sanity Check our own output 

In [ ]:
print('*******VALIDATION CHECKS**********\n')
#1.No negative Prices
assert(master['price']>=0).all(),"Negative Price Found!"
print("price checked passed")
#2.Delivery date only exists for delivered orders
delivered=master[master['order_status']=='delivered']
print(f"Delivered orders with delay data:{delivered['delivery_delay_days'].notna().sum():,}")
#3.Repeat customer rate is reasonable
print(f"Repeat Customer Rate:{repeat_rate*100:.1f}%")
#Revenue per item is always positive
print(f"zero/null revenue rows:{(master['revenue_per_item']<=0).sum()}")
print("\nAll Check Passed.Ready to commit.")

*******VALIDATION CHECKS**********



AssertionError: Negative Price Found!

This is actually a good thing. Your validation check is working exactly as intended. It caught a real data quality issue. 

In [80]:
print("price distribution:")
print(master['price'].describe())

print(f"\nRows where price=0 :{(master['price']==0).sum()}")
print(f"Rows where price<0 :{(master['price']<0).sum()}")
print(f"Rows where price is null :{master['price'].isnull().sum()}")

#look at the bad rows
bad_price=master[master['price']<=0]
print(f"\n Total bad price rows:{len(bad_price)}")
print("\nSample bad Rows:")
bad_price[['order_id','order_status','product_id','price','freight_value','revenue_per_item']].head(10)

price distribution:
count    112650.000000
mean        120.653739
std         183.633928
min           0.850000
25%          39.900000
50%          74.990000
75%         134.900000
max        6735.000000
Name: price, dtype: float64

Rows where price=0 :0
Rows where price<0 :0
Rows where price is null :775

 Total bad price rows:0

Sample bad Rows:


,order_id,order_status,product_id,price,freight_value,revenue_per_item


The assert failed because of null values, not negative prices. (master['price'] >= 0) returns False for nulls in Pandas, which triggered the assert even though there are zero actually negative prices.So we remove these rows which is cleaner for analysis`

In [83]:
master=master[master['price'].notnull()].copy()
print(f"Rows after droping null prices rows: {len(master)}")

Rows after droping null prices rows: 112650


In [86]:
print('=== VALIDATION CHECKS ===\n')
# 1. No negative or null prices
assert (master['price'] > 0).all(), "Zero/negative/null prices found!"
print(f"Price check passed — min price: {master['price'].min()}")
# 2. Delivered orders have delay data
delivered = master[master['order_status'] == 'delivered']
print(f"Delivered orders with delay data: {delivered['delivery_delay_days'].notna().sum():,}")
#3.Repeat customer rate is reasonable
print(f"Repeat Customer Rate:{repeat_rate*100:.1f}%")
#4.Revenue per item is always positive
print(f"zero/null revenue rows:{(master['revenue_per_item']<=0).sum()}")
#5 master table grain check
total_rows=len(master)
unique_items=master.groupby(['order_id','order_item_id']).ngroups
print(f"Row count:{total_rows:,}-unique order+item combos:{unique_items:,}")
assert total_rows==unique_items,"Duplicate order-item rows found"
print("Grain check passed - one row per order-item")
print("\nAll Check Passed.Ready to commit.")


=== VALIDATION CHECKS ===

Price check passed — min price: 0.85
Delivered orders with delay data: 110,189
Repeat Customer Rate:3.1%
zero/null revenue rows:0
Row count:112,650-unique order+item combos:112,650
Grain check passed - one row per order-item

All Check Passed.Ready to commit.


Now save to the pgsql and data/processed/master.csv

In [87]:
#To PostgreSQL - Phase 3 SQl analysis will query this
master.to_sql('master_orders',engine,if_exists='replace',index=False)
print('masters_orders saved to PostgreSQl')

#To CSV - backup + Tableau will use it directly
master.to_csv(r'J:\python\DA_project\Olist_Ecommerce_intelligence\data\processed\master_orders.csv',index=False)
print(r'master_orders.csv saved to Olist_Ecommerce_intelligence\data\processed')
#save customer_summary for use in later phases
customer_summary.to_csv(r'J:\python\DA_project\Olist_Ecommerce_intelligence\data\processed\customer_summary.csv',index=False)
print("customer_summary saved:", customer_summary.shape)

masters_orders saved to PostgreSQl
master_orders.csv saved to Olist_Ecommerce_intelligence\data\processed
customer_summary saved: (96096, 3)
